### PR TITLE
Conrib zsr/python logging

### PR DESCRIPTION
**Pull request by zyd14**
Adds an introduction to the powerful python logging module.  Any feedback is highly welcome!

### CODE REVIEW

**Comment by sminot** [comment]
As someone who uses the `logging` module of Python quite frequently, I can say that these docs appear to be extremely well written and useful. Great job!

File: _compdemos/python_logging_demo.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+focused log streams that can track logs from across multiple services if they all use `WatchTower` to point to the same log
+group / stream. I've found `WatchTower` helpful for consolidating Lambda logs as by default a new stream is created for 
+every new lambda instance, which can get really confusing, particularly if you have concurrent executions.  I have had 
+some issues (which may have more to do with how I'm creating the `WatchTower` handlers and my application structure) 
+with some of the functions hitting AWS service limits on API calls like Create/DescribeLogGroups and Create/DescribeLogStreams, 
+which can have have an affect on responsiveness and at worst can crash the Lambda call if the exceptions aren't handled properly. 
+This is particularly the case when the service is under a high load. If someone has another solution for this I'd love to hear about it!
+
+
+Give me a shout at `zromer at fredhutch.org` if you have any questions, comments, or corrections you think should be made.
```
Comment (WhiteAu) [inline _compdemos/python_logging_demo.md:236]: Thank you for making this! 
Thoughts: either a separate page on logging patterns/a subsection with some quick examples of logging patterns like
main.py: 
> logger = logging.getLogger() #get's the Base Logger, which everything inherits from
> logger.basicConfig(...) #setting this on the Base Logger is inheritable on every other logger that implicitly inherits from it
important_module.py: 
> logger = logging.getLogger(__name__) #this automagically sets the logger to the name of the module (ie: 'important_module.py')
> #additionally, all BasicConfig is inherited from the logger in main.py at this point

File: _compdemos/python_logging_demo.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+logger = logging.getLogger('new_logger')
+logger.setLevel(logging.DEBUG)
+file_formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s - line %(lineno)s - module %(module)s - function %(funcName)s - file %(fileName)s')
+file_handler = logging.FileHandler(filename='verbose_logger.log')
+file_handler.setFormatter(file_formatter)
+
+slim_formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s - line %(lineno)s')
+slim_console_handler = logging.StreamHandler()
+slim_console_handler.setFormatter(slim_formatter)
+logger.addHandler(slim_console_handler)
```
Comment (WhiteAu) [inline _compdemos/python_logging_demo.md:169]: worth adding an inline comment at this point emphasizing that adding multiple handlers means that the logger will write to console AND write to file at this point?

File: _compdemos/python_logging_demo.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+logger.addHandler(file_handler)
+```
+
+### Filters 
+`Filters` are attached to `Handlers` and provided filtering of messages from below a specified point in the logger hierarchy
```
Comment (WhiteAu) [inline _compdemos/python_logging_demo.md:174]: this presently isn't shown above in the code snippet: the filter is being applied at the logger level-- perhaps a second code snippet adding two FileHandlers: infoLogs and errorLogs with different message levels specified?

File: _compdemos/python_logging_demo.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+where the parent-child relationships of our loggers can be described as `base_logger.module_logger` and `base_logger.debug_logger`. 
+A `Filter` object can be attached to a `Logger` instance in order to prevent it from propagating messages from specific parts of the 
+logger hierarchy. `Filter` objects can also be attached to specific handlers as well, in order for more nuanced control of which loggers 
+are emitted to a particular output location.  
+For instance, if we were to want to have messages from our `module_logger` propagate to the `base_logger`, but did not want any messages 
+from `debug_logger` propagating up to the `base_logger` and its handlers, a `Filter` object could be attached to the `base_logger`
+specifying that it should ignore any messages recorded by `debug_logger`   
+
+### Log levels  
+Both `Logger` and `Handler` instances have an attribute called `.level`, which determines the severity of messages the
```
Comment (WhiteAu) [inline _compdemos/python_logging_demo.md:191]: his presently isn't shown above in the code snippet: the filter is being applied at the logger level-- perhaps a second code snippet adding two FileHandlers: infoLogs and errorLogs with different message levels specified?

File: _compdemos/python_logging_demo.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+It is also possible to create your own log levels, a process which is described in the Logging Cookbook.
+
+## Where/When to Implement Logging
+I like to log early and log often, especially during the development phase of a project.  In particular, I find often 
+find myself logging the responses to remote calls so I can always have a reference in my log of what the response looked
+like before I start processing it.  When writing REST APIs I often will log the raw request prior to any processing, so
+that when something goes wrong I can look at what data I started with and investigate from there. I've found logging to 
+be invaluable for debugging stateless applications in AWS, as they pick up log messages emitted to STDOUT from your program so I 
+can log state at various points in the application to better understand what might have gone wrong.  In most of my projects I
+create a function to create or retrieve a pre-configured `Logger` instance early on, and use it constantly. I then use
```
Comment (WhiteAu) [inline _compdemos/python_logging_demo.md:213]: I like this! Perhaps a code snippet example?

File: _compdemos/python_logging_demo.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+module level, meaning all logging calls below that won't get reported.  As an example you might want a bunch of `.debug()`
+calls around a particularly tricky piece of logic when running in your test environment, but perhaps want to ignore `.debug()`
+and only see`.error()` and above messages in another environment.  You could have a LOG_LEVEL environment variable which 
+is then used in the code to set your logger levels to their desired state, without having to touch the code. Example below:  
+
+```python
+import logging
+import os
+
+LOG_LEVEL = os.getenv('LOG_LEVEL')
```
Comment (WhiteAu) [inline _compdemos/python_logging_demo.md:109]: consider changing to
> os.getenv('LOG_LEVEL', logging.INFO) to emphasize a fallback value?

File: _compdemos/python_logging_demo.md (hunk 1 above)
Comment (esilgard) [reply _compdemos/python_logging_demo.md:236]: +1 thanks for contributing this

**Comment by vortexing** [comment]
So @zyd14 are we good?  Think we should merge and delete the branch??

File: _compdemos/python_logging_demo.md (hunk 6 above)
Comment (zyd14) [reply _compdemos/python_logging_demo.md:109]: Great idea!

File: _compdemos/python_logging_demo.md (hunk 1 above)
Comment (zyd14) [reply _compdemos/python_logging_demo.md:236]: I could definitely whip up some code examples for functions that configure different handlers / levels / filters for loggers that someone could just drag and drop into their projects.

File: _compdemos/python_logging_demo.md (hunk 4 above)
Comment (zyd14) [reply _compdemos/python_logging_demo.md:191]: A simple example of two different fileHandlers with different log levels might be a great way to explain log levels.  I'll consider adding a code snippet for that.  Thanks!

**Pull request merged**